---
# Please do not edit this file directly; it is auto generated.
# Instead, please edit 02-Experimental-Considerations.md in _episodes_rmd/
source: Rmd
title: "Experimental Considerations"
teaching: 10
exercises: 2
questions:
- "How do I design a single cell RNAseq experiment?"
objectives:
- "Understand the importance of biological replication for reproducibility and rigor."
- "Understand the differences between pooling methods and select between them."
- "Understand why confounding experimental batch with the variable of interest makes it impossible to disentangle the two."
- "Understand the different data modalities in single-cell sequencing and be able to select a modality to answer an experimental question."
keypoints:
- "Edit the .Rmd files not the .md files"
- "Run `make serve` to knit documents and preview lesson website locally"
---



## Biological replication  



## Pooling options  



### Avoid Confounding Batch with Experimental Variables

When designing an experiment, it is tempting to lay samples out in some order that is easy to remember. If you are performing a dose-response experiment, you might order your samples from the lowest dose to the highest dose. If you are assaying transcript levels in different tissues, you might sort the samples by tissue. However, if you maintain this order, you risk confounding RNA-seq batches with your experimental batches. This is called "confounding" and it makes it impossible to statistically disentangle the batch effect from your experimental question.

Let's look at an example. Suppose that you have analyzed cells from kidney, liver, spleen, and heart. In the lab, you would process each tissue type independently because each tissue is distinct. If you maintain the tissues separately as you perform library preparation and sequencing, you will be unable to determine whether the differences that you observe are due to tissue differences or RNA-seq batch differences. 

In order to reduce the effect of batch structure on your experiment is it crucial to employ **randomization**. This involves randomizing tissues across library preparation and sequencing batches. The figure below shows examples of poor and good batch design. In the left panel, the library from each tissue is placed in a separate flow cell. This means that, if there are differences between flow cells, there will be no way to distinguish or correct these differences. On the right, we show the recommended flow cell design. The library from each tissue should be distributed evenly across each flow cell. The reads for each tissue can be combined from each flow cell using the cell bar codes after sequencing.

![Batch Confounding](../fig/flow_cell_batch.png)


## Different data modalities (cell vs nuclei, RNA vs ATAC, etc) 


## Brief overview of choices available from 10X for how to run samples

> ## Recommended Reading
>
> - Review(s) with best-practices or advice on experimental design?
>
{: .checklist }

### Session Info


~~~
sessionInfo()
~~~
{: .language-r}



~~~
R version 4.1.2 (2021-11-01)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19042)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252 
[2] LC_CTYPE=English_United States.1252   
[3] LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C                          
[5] LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] knitr_1.39

loaded via a namespace (and not attached):
[1] compiler_4.1.2 magrittr_2.0.3 tools_4.1.2    stringi_1.7.6  stringr_1.4.0 
[6] xfun_0.31      evaluate_0.15 
~~~
{: .output}
