---
# Please do not edit this file directly; it is auto generated.
# Instead, please edit 05-Common-Analyses.md in _episodes_rmd/
source: Rmd
title: "Common Analyses"
teaching: 10
exercises: 2
questions:
- "What are the most common single cell RNAseq analyses?"
objectives:
- "Explain how to use RMarkdown with the new lesson template."
- "Demonstrate how to include pieces of code, figures, and challenges."
keypoints:
- "Edit the .Rmd files not the .md files"
- "Run `make serve` to knit documents and preview lesson website locally"
---



## Read Data from Previous Lesson


~~~
liver_iv = readRDS(file.path(data_dir, 'lesson04.rds'))
~~~
{: .language-r}



~~~
Warning in gzfile(file, "rb"): cannot open compressed file '../data/
lesson04.rds', probable reason 'No such file or directory'
~~~
{: .warning}



~~~
Error in gzfile(file, "rb"): cannot open the connection
~~~
{: .error}


## Normalization (log and more specialized) 

> TBD: How do we show a histogram of the unnormalized and normalized values?

The count data is usually log-normally distributed. Many statistical methods work best when the data is normally distributed. 


~~~
liver_iv <- liver_iv %>%
              NormalizeData(normalization.method = "LogNormalize")
~~~
{: .language-r}



~~~
Error in NormalizeData(., normalization.method = "LogNormalize"): object 'liver_iv' not found
~~~
{: .error}


~~~
liver_iv <- liver_iv %>% 
              FindVariableFeatures(nfeatures = 2000)
~~~
{: .language-r}



~~~
Error in FindVariableFeatures(., nfeatures = 2000): object 'liver_iv' not found
~~~
{: .error}



~~~
liver_iv <- liver_iv %>%
              ScaleData(vars.to.regress = c("percent.mt", "nCount_RNA"))
~~~
{: .language-r}



~~~
Error in ScaleData(., vars.to.regress = c("percent.mt", "nCount_RNA")): object 'liver_iv' not found
~~~
{: .error}


~~~
liver_iv <- liver_iv %>%
              RunPCA(verbose = FALSE, npcs = 100)
~~~
{: .language-r}



~~~
Error in RunPCA(., verbose = FALSE, npcs = 100): object 'liver_iv' not found
~~~
{: .error}
## Dimensionality reduction (UMAP, tSNE, etc) 

## Clustering 


~~~
ElbowPlot(liver_iv, ndims = 100) + geom_vline(xintercept = 27)
~~~
{: .language-r}



~~~
Error in Stdev(object = object, reduction = reduction): object 'liver_iv' not found
~~~
{: .error}



~~~
num_pc <- 27
liver <- FindNeighbors(liver_iv, reduction = 'pca', 
                       dims = 1:num_pc, verbose = FALSE) %>%
           FindClusters(verbose = FALSE, resolution = 0.8) %>%
           RunUMAP(reduction = 'pca', dims = 1:num_pc, verbose = FALSE)
~~~
{: .language-r}



~~~
Error in FindNeighbors(liver_iv, reduction = "pca", dims = 1:num_pc, verbose = FALSE): object 'liver_iv' not found
~~~
{: .error}

## Annotating cell types (+ automated options e.g. SingleR) 

## Finding marker genes 

## Discussion of data visualization, sample of common plots

