---
# Please do not edit this file directly; it is auto generated.
# Instead, please edit 06-Biology-Driven-Analyses.md in _episodes_rmd/
source: Rmd
title: "Biology Driven Analyses of scRNA-Seq"
teaching: 10
exercises: 2
questions:
- "What are some scRNA-Seq analyses that might provide me with biological insight?"
objectives:
- "Understand the ability (and limitations) of scRNA-Seq data for quantifying differences in gene expression and tissue cell composition."
- "Be able to articulate the concept of pseudotime."
- "Have basic ability to be able to conduct enrichment analyses of gene expression and look at how cells interact in scRNA-Seq"
keypoints:
- "ADD"
- "ADD"
---



## Read Data from Previous Lesson


~~~
liver <- load(file.path(data_dir, 'lesson05.Rdata'))
~~~
{: .language-r}



~~~
Error in file.path(data_dir, "lesson05.Rdata"): object 'data_dir' not found
~~~
{: .error}

## Identifying cell types

Let's plot the expression of some of the major cell type
markers. Look at the data.frame `markers` for a summary of the
markers we found in lesson 5.


~~~
UMAPPlot(liver, label = TRUE, label.size = 6) + NoLegend()
~~~
{: .language-r}



~~~
Error in UMAPPlot(liver, label = TRUE, label.size = 6): could not find function "UMAPPlot"
~~~
{: .error}



~~~
FeaturePlot(liver, c("Lyve1", "Cdh5"), cols = c('lightgrey', 'red'))
~~~
{: .language-r}



~~~
Error in FeaturePlot(liver, c("Lyve1", "Cdh5"), cols = c("lightgrey", : could not find function "FeaturePlot"
~~~
{: .error}

## Differential expression 

Looking for differential expression can be thought of as a problem that
is related to finding cell type marker genes. Marker genes are, by definition,
genes that vary significantly between cell types. Often we are most interested
in the expression of genes that specifically mark particular cell types 
that we are interested in, but there can also be value in using broader
markers (e.g. CD45 - encoded by the gene *Ptprc* - marks all immune cells).

<!-- What do we look for differential expression between??? -->

## Pathway enrichment 

We could conduct pathway enrichment on 

## Differences in cell composition

Single cell transcriptomics is a very powerful method for identifying
differences in tissue cell composition. This sort of question was 
usually addressed using methods like flow cytometry in the past, but flow
cytometry requires knowing what you want to look for before you can
quantify it. In scRNA-Seq we can take an entirely unbiased view of cell
composition, and we can define cell subsets with more subtle changes in
gene expression as opposed to only looking at broad classes of cell types.


scCODA
linear modeling
propeller

## Cell-cell communication 

cellchat, cellphoneDB, etc.

## Pseudotime and velocity

What is pseudotime

There are many, many algorithms and software packages for inferring 
pseudotime. For a very nice list check out
[Anthony Gitter's github list](https://github.com/agitter/single-cell-pseudotime).

What is RNA velocity?

monocle
RNA velocity
scvelo


