---
source: Rmd
title: "Experimental Considerations"
teaching: 10
exercises: 2
questions:
- "How do I design a rigorous and reproducible single cell RNAseq experiment?"
objectives:
- "Understand the importance of biological replication for rigor and reproducibility."
- "Understand the differences between pooling methods and choose between them."
- "Understand why confounding experimental batch with the variable of interest makes it impossible to disentangle the two."
- "Understand the different data modalities in single-cell sequencing and be able to select a modality to answer an experimental question."
keypoints:
- "Due to the increased variance in single-cell data sets, a well-powered study with adequate biological replication is essential for rigor & reproducibility."
- "Do not confound experimental batch with any technical aspect of the experiment, i.e. sample pooling or flow cell assignment."
- "Different experimental questions can be investigated using different single-cell sequencing modalities."
---

```{r, include=FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("02-")
```

## Biological replication  

> TBD: DAS to fill in.
> TBD: Is there a way to do power analyses for scRNA-seq? For number of cell types? For DE genes?

## Pooling options  

> TBD: DAS to fill in.

1. Genetic demultiplexing: Mix cells from different strains in same tube and use genotyping information to demultiplex cells.
2. Cell-hashing: use barcode-carrying antibody that binds to universal cell surface marker.

> How to decide on sequencing depth in experiment?

### Avoid Confounding Batch with Experimental Variables

When designing an experiment, it is tempting to lay samples out in some order that is easy to remember. If you are performing a dose-response experiment, you might order your samples from the lowest dose to the highest dose. If you are assaying transcript levels in different tissues, you might sort the samples by tissue. However, if you maintain this order, you risk confounding RNA-seq batches with your experimental batches. This is called "confounding" and it makes it impossible to statistically disentangle the batch effect from your experimental question.

Let's look at an example. Suppose that you have analyzed cells from kidney, liver, spleen, and heart. In the lab, you would process each tissue type independently because each tissue is distinct. If you maintain the tissues separately as you perform library preparation and sequencing, you will be unable to determine whether the differences that you observe are due to tissue differences or RNA-seq batch differences. 

In order to reduce the effect of batch structure on your experiment is it crucial to employ **randomization**. This involves randomizing tissues across library preparation and sequencing batches. The figure below shows examples of poor and good batch design. In the left panel, the library from each tissue is placed in a separate flow cell. This means that, if there are differences between flow cells, there will be no way to distinguish or correct these differences. On the right, we show the recommended flow cell design. The library from each tissue should be distributed evenly across each flow cell. The reads for each tissue can be combined from each flow cell using the cell bar codes after sequencing.

![Batch Confounding](../fig/flow_cell_batch.png)


## Different data modalities (cell vs nuclei, RNA vs ATAC, etc) 

> TBD: DAS to fill in.

* RNA-seq
* ATAC-seq
* Immune repertoire seq: clonality of adaptive immunne cells.
* cite-seq: cell surface protein.

> ## Challenge 1
> TBD: Can we propose an experimental question that naturally leads to choosing one of the modalities?
>
> > ## Solution to Challenge 1
> >
> >    
> {: .solution}
{: .challenge}

## Brief overview of choices available from 10X for how to run samples

> TBD: Discuss full-transcript vs 3' sequencing here?

> TBD: DAS to fill in.

> ## Recommended Reading
>
> - Review(s) with best-practices or advice on experimental design?
>
{: .checklist }

### Session Info

```{r session_info}
sessionInfo()
```
