---
source: Rmd
title: "Overview of scRNA-seq Data"
teaching: 10
exercises: 2
questions:
- "What does single cell RNAseq data look like?"
objectives:
- "Explain how to use RMarkdown with the new lesson template."
- "Demonstrate how to include pieces of code, figures, and challenges."
keypoints:
- "Edit the .Rmd files not the .md files"
- "Run `make serve` to knit documents and preview lesson website locally"
---

```{r, include=FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("03-")
```

## What files are delivered? Details of FASTQ, etc. 

## Typical pre-processing pipeline -- 10X CellRanger 

## Intro to two major single cell ecosystems: 

* R/Seurat : major focus of this workshop. Biggest strength is straightforward vignettes and ease of visualization/exploration 
* Python/scanpy

