---
source: Rmd
title: "Quality Control of scRNA-Seq Data"
teaching: 90
exercises: 30
questions:
- "How do I determine if my single cell RNA-seq experiment data is high quality?"
- "What are the common quality control metrics that I should check in my scRNA-seq data?"
objectives:
- "Critically examine scRNA-seq data to identify potential technical issues."
- "Apply filters to remove cells that are largely poor quality/dead cells."
- "Understand the implications of different filtering steps on the data."
keypoints:
- "It is essential to filter based on criteria including mitochondrial gene expression and number of genes expressed in a cell."
- "Determining your filtering thresholds should be done separately for each experiment, and these values can vary dramatically in different settings."
---

```{r, setup, include=FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("04-")
```

```{r libs, warning=FALSE}
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(Matrix))
suppressPackageStartupMessages(library(SingleCellExperiment))
suppressPackageStartupMessages(library(scds))
suppressPackageStartupMessages(library(Seurat))

data_dir <- '../data'
```


```{r load_data,include=FALSE}
# Only needed to build lessons.
# Don't use in course.
load(file.path(data_dir, 'lesson03.Rdata'))
```

> DMG: Overall process figure.

## Quality control in scRNA-seq

There are many technical reasons why cells produced by an scRNA-seq protocol 
might not be of high quality. The goal of the quality control steps are to 
assure that only single, live cells are included in the final data set. 
Ultimately some multiplets and poor quality cells will likely escape your
detection and make it into your final dataset; however, these quality
control steps aim to reduce the chance of this happening.
Failure to undertake quality control is likely to adversely impact cell type 
identification, clustering, and interpretation of the data. 

Some technical questions that you might ask include:

1. Why is mitochondrial gene expression high in some cells?
1. What is a unique molecular identifier (UMI), and why do we check numbers of UMI?
1. What happens to make gene counts low in a cell?

> DMG: Add image for UMI? Too complicated? Yes.

<!-- DASremoved ribosomal point here. Often we don't filter based on ribo, not clearly associated with poor quality cells -->

## Doublet detection 

We will begin by discussing doublets. We have already discussed the 
concept of the doublet. Now we will try running one computational
doublet-detection approach and track predictions of doublets.

We will use the scds method. scds contains two methods for predicting
doublets. Method cxds is based on co-expression of gene pairs, while 
method bcds uses the full count information and a binary classification 
approach using in silico doublets. 
Method cxds_bcds_hybrid combines both approaches. We will use the combined
approach. See 
[Bais and Kostka 2020](https://academic.oup.com/bioinformatics/article/36/4/1150/5566507)
for more details.

Because this doublet prediction method takes some time and is
a bit memory-intensive, we will run it only on cells from one mouse.
We will return to the doublet predictions later in this lesson.


```{r scds1}
cell_ids <- filter(metadata, sample == 'CS52') %>% pull(cell)
sce      <- SingleCellExperiment(list(counts = counts[, cell_ids]))
sce      <- cxds_bcds_hybrid(sce)
doublet_preds <- colData(sce)
```

```{r gc, echo = FALSE, message = FALSE}
rm(sce)
gc()
```


## High-level overview of quality control and filtering

First we will walk through some of the typical quantities one 
examines when conducting quality control of scRNA-Seq data.


### Filtering Genes by Counts

As mentioned in an earlier lesson, the counts matrix is sparse and may contain 
rows (genes) or columns (cells) with low overall counts. In the case of genes, 
we likely wish to exclude genes with zeros counts in most cells. Let's see how 
many genes have zeros counts across all cells. Note that the 
[Matrix package](https://cran.r-project.org/web/packages/Matrix/index.html) 
has a special implementation of 
[rowSums](https://rdrr.io/rforge/Matrix/man/colSums.html) which works with 
sparse matrices.

```{r zero_gene_counts}
gene_counts <- Matrix::rowSums(counts)
sum(gene_counts == 0)
```

> DMG: Figure for row-summing???

Of the `r nrow(counts)` genes, `r sum(gene_counts == 0)` have zero counts across 
all cells. These genes do not inform us about the mean, variance, or covariance 
of any of the other genes and we will remove them before proceeding with 
further analysis.

```{r filter_gene_by_counts}
counts <- counts[gene_counts > 0,]
```

This leaves `r nrow(counts)` genes in the counts matrix.

We could also set some other threshold for filtering genes. Perhaps we should 
look at the number of genes that have different numbers of counts. We will use 
a histogram to look at the distribution of overall gene counts. Note that, since 
we just resized the counts matrix, we need to recalculate `gene_counts`.

We will count the number of cells in which each gene was detected. Because 
`counts` is a sparse matrix, we have to be careful not to perform operations 
that would convert the entire matrix into a non-sparse matrix. This might 
happen if we wrote code like:

```{}
gene_counts <- rowSums(counts > 0)
```

The expression `counts > 0` would create a logical matrix that takes up much 
more memory than the sparse matrix. We might be tempted to try 
`rowSums(counts == 0)`, but this would also result in a non-sparse matrix 
because most of the values would be `TRUE`. However, if we evaluate 
`rowSums(counts != 0)`, then most of the values would be `FALSE`, which can be 
stored as 0 and so the matrix would still be sparse. The `Matrix` package has 
an implementation of 'rowSums()' that is efficient, but you may have to specify 
that you want to used the `Matrix` version of 'rowSums()' explicitly.

```{r gene_count_hist}
gene_counts <- tibble(counts  = Matrix::rowSums(counts > 0))

gene_counts %>% 
  ggplot(aes(counts)) +
    geom_histogram(bins = 1000) +
    labs(title = 'Histogram of Number of Cells in which Gene was Detected',
         x     = 'Number of Genes',
         y     = 'Number of Cells in which Gene was Detected') +
  theme_bw(base_size = 14)
```

As you can see, the number of cells in which each gene is detected spans several 
orders of magnitude and this makes it difficult to interpret the plot. Some 
genes are detected in all cells while others are detected in only one cell. 
Let's zoom in on the part with lower gene counts.

```{r gene_count_hist_2}
gene_counts %>%
  ggplot(aes(counts)) +
    geom_histogram(bins = 1000) +
    labs(title = 'Histogram of Number of Cells in which Gene was Detected',
         x     = 'Number of Genes',
         y     = 'Number of Cells in which Gene was Detected') +
    lims(x = c(0, 50)) +
    theme_bw(base_size = 14) +
    annotate('text', x = 2, y = 1500, hjust = 0,
             label = str_c(sum(gene_counts == 1), ' genes were detected in only one cell')) +
    annotate('text', x = 3, y = 900, hjust = 0,
             label = str_c(sum(gene_counts == 2), ' genes were detected in two cells'))
```

In the plot above, we can see that there are `r sum(gene_counts == 1)` 
genes that were detected in only one cell, `r sum(gene_counts == 2)` 
genes detected in two cells, etc.

Making a decision to keep or remove a gene based on its expression being
detected in a certain
number of cells can be tricky. 
If you retain all genes, you may consume more computational resources and add 
noise to your analysis. If you discard too many genes, you may miss rare 
but important cell types.

Consider this example: You have a total of 10,000 cells in your scRNA-seq 
results. There is a rare cell population consisting of 100 cells that 
expresses 20 genes which are not expressed in any other cell type. If you 
only retain genes that are detected in more than 100 cells, you will miss 
this cell population.


> ## Challenge 1
> How would filtering genes too strictly affect your results? 
How would this affect your ability to discriminate between cell types? 
>
> > ## Solution to Challenge 1
> > 
> > Filtering too strictly would make it more difficult to distinguish between
cell types. The degree to which this problem affects your analyses depends on
the degree of strictness of your filtering. Let's take the situation to its
logical extreme -- what if we keep only genes expressed in at least 95% of cells.
If we did this, we would end up with only `r sum(gene_counts$counts > ncol(counts) * 0.95)`
genes! By definition these genes will be highly expressed in all cell types,
therefore eliminating our ability to clearly distinguish between cell types.
> {: .solution}
{: .challenge}


> ## Challenge 2
> What total count threshold would you choose to filter genes? Remember that 
there are `r ncol(counts)` cells.
>
> > ## Solution to Challenge 2
> >
> > This is a question that has a somewhat imprecise answer. Following from
challenge one, we do not want to be *too* strict in our filtering. However,
we do want to remove genes that will not provide much information about
gene expression variability among the cells in our dataset. Our recommendation
would be to filter genes expressed in < 5 cells, but one could reasonably
justify a threshold between, say, 3 and 20 cells.
> {: .solution}
{: .challenge}

> DAS & DMG: Should we have some kind of challenge later in the lesson in which the students do no filtering vs. filtering by genes that occur in 100+ cells? Should the challenge be later in the clusering lesson? Filter more strictly and redo analysis?

### Filtering Cells by Counts

Next we will look at the number of genes expressed in each cell. 
If a cell lyses and leaks RNA,the total number of reads in a 
cell may be low, which leads to lower gene counts. Furthermore, each
single cell suspension likely contains some amount of so-called "ambient"
RNA from damaged/dead/dying cells. This ambient RNA comes along for the ride
in every droplet. Therefore even droplets that do not contain cells
(empty droplets) can have some reads mapping to transcripts that look
like gene expression.
Filtering out these kinds of cells is a quality control step that 
should improve your final results.

We will explicitly use the `Matrix` package's implementation of 'colSums()'.

```{r sum_cell_counts}
tibble(counts  = Matrix::colSums(counts > 0)) %>%
  ggplot(aes(counts)) +
    geom_histogram(bins = 500) +
    labs(title = 'Histogram of Number of Genes per Cell',
         x     = 'Number of Genes with Counts > 0',
         y     = 'Number of Cells')
```

Cells with way more genes expressed than the typical cell might be
doublets/multiplets and should also be removed.

### Creating the Seurat Object

<img src="../fig/single_cell_flowchart_2.png" width="800px" alt="Single Cell Flowchart" >

In order to use Seurat, we must take the sample metadata and gene counts and 
create a 
[Seurat Object](https://rdrr.io/cran/SeuratObject/man/Seurat-class.html). 
This is a data structure which organizes the data and metadata and will 
store aspects of the analysis as we progress through the workshop.

Below, we will create a Seurat object for the liver data. We must first 
convert the cell metadata into a data.frame and place the barcodes 
in rownames. The we will pass the counts and metadata into the
[CreateSeuratObject](https://search.r-project.org/CRAN/refmans/SeuratObject/html/CreateSeuratObject.html) 
function to create the Seurat object. 

In the section above, we examined the counts across genes and cells and 
proposed filtering using thresholds.  The CreateSeuratObject function 
contains two arguments, 'min.cells' and 'min.features', that allow us to 
filter the genes and cells by counts. Although we might use these arguments
for convenience in a typical analysis, for this course we will look more
closely at these quantities on a per-library basis to decide on
our filtering thresholds. We will us the 'min.cells' argument to filter out
genes that occur in less than 5 cells.

```{r seed}
# set a seed for reproducibility in case any randomness used below
set.seed(1418)
```

```{r create_seurat_obj}
metadata <- as.data.frame(metadata) %>%
              column_to_rownames('cell')
liver <- CreateSeuratObject(counts    = counts, 
                            project   = 'liver: scRNA-Seq',
                            meta.data = metadata,
                            min.cells = 5)
```

We now have a Seurat object with `r nrow(liver)` genes and `r ncol(liver)` cells.

We will remove the counts object to save some memory because it is now stored 
inside of the Seurat object.

```{r rm_counts}
rm(counts)
gc()
```


Add on doublet predictions that we did earlier in this lesson.

```{r add_doublets}
liver <- AddMetaData(liver, as.data.frame(doublet_preds))
```

Let's briefly look at the structure of the Seurat object. The counts are stored 
as an [assay](https://github.com/satijalab/seurat/wiki/Assay), which we can query using the `Assays()` function. 

```{r get_assays}
Seurat::Assays(liver)
```

The output of this function tells us that we have data in an "RNA assay. We can access this using the [GetAssayData](https://mojaveazure.github.io/seurat-object/reference/AssayData.html) function.

```{r get_assay_data}
tmp = GetAssayData(object = liver, slot = 'data')
tmp[1:5,1:5]
```

As you can see the data that we retrieved is a sparse matrix, just like the counts that we provided to the Seurat object.

What about the metadata? We can access the metadata to using somewhat confusing double square bracket syntax.

```{r show_meta}
head(liver[[]])
```

Notice that there are some columns that were not in our original metadata file; 
specifically the 'nCount_RNA' and 'nFeature_RNA' columns. 

* *nCount_RNA* is the total counts for each cell. 
* *nFeature_RNA* is the number of genes with counts > 0 in each cell.

These were calculated by Seurat when the Seurat object was created. We will use 
these later in the lesson.


## Typical filters for cell quality

Here we briefly review these filters and decide what thresholds we will
use for these data.


### Filtering by Mitochondrial Gene Content

<img src="../fig/single_cell_flowchart_3.png" width="800px" alt="Single Cell Flowchart" >

During apoptosis, the cell membrane may break and release transcripts into 
the surrounding media. However, the mitochondrial transcripts may remain inside 
of the mitochondria. This will lead to an apparent, but spurious, increase in 
mitochondrial gene expression. As a result, we use the percentage of 
mitochondrial-encoded reads to filter out cells that were not healthy during 
sample processing. See 
[this link](https://kb.10xgenomics.com/hc/en-us/articles/360001086611) 
from 10X Genomics for additional information.

First we compute the percentage mitochondrial gene expression in each cell.

```{r pct_mito}
liver <- liver %>% 
              PercentageFeatureSet(pattern = "^mt-", col.name = "percent.mt")
```

Different cell types may have different levels of mitochondrial RNA content.
Therefore we must use our knowledge of the particular biological system
that we are profiling in order to choose an appropriate threshold.
If we are profiling single nuclei instead of single cells we might 
consider a very low threshold for MT content. If we are profiling a tissue
where we anticipate broad variability in levels of mitochondrial RNA
content between cell types, we might use a very lenient threshold
to start and then return to filter out additional cells after we 
obtain tentative cell type labels that we have obtained by carrying
out normalization and clustering. In this course we will filter only once


```{r seurat_counts_plots}
VlnPlot(liver, features = "percent.mt", group.by = 'sample')
```

It is hard to see with so many dots! Let's try another version where we just
plot the violins:

```{r seurat_counts_plots2,fig.height=4,fig.width=6}
VlnPlot(liver, features = "percent.mt", group.by = 'sample', pt.size = 0)
```

Library "CS89" (and maybe CS144) have a "long tail" of cells with high
mitochondrial
gene expression. We may wish to monitor these libraries throughout QC
and decide whether it has problems worth ditching the sample.

```{r mito_by_cell_type, include = FALSE, eval = FALSE}
# DMG made the plot below to see if there are mitochondrial expression differences by annotated cell type. The students won't have this file at this stage of the analysis. But how do we discuss these differences? Since we're looking for high values, it may not be too important.
liver[[c('annot', 'percent.mt')]] %>% 
    ggplot(aes(annot, percent.mt + 0.01)) +
      geom_boxplot() +
      scale_y_log10() +
      coord_flip()
```

In most cases it would be ideal to determine separate filtering 
thresholds on each scRNA-Seq sample. This would account for the fact
that the characteristics of each sample might vary 
(for many possible reasons) even if the
same tissue is profiled. However, in this course we will see if we can
find a single threshold that works decently well across all samples.
As you can see, the samples we are examining do not look 
drastically different so this may not be such an unrealistic simplification.

We will use a threshold of 14% mitochondrial gene expression which will 
remove the "long tail" of cells with high `percent.mt` values. We could
also perhaps justify going as low as 10% to be more conservative,
but we likely would not want to go down to 5%, which would 
remove around half the cells.

```{r subset_by mito}
#liver <- subset(liver, subset = percent.mt < 14)
```

### Filtering Cells by Total Gene Counts

Let's look at how many genes are expressed in each cell. 
Again we'll split by
the mouse ID so we can see if there are particular samples that are 
very different from the rest.
Again we will show only the violins for clarity.

```{r filter_gene_counts,fig.height=4,fig.width=6}
VlnPlot(liver, 'nFeature_RNA', group.by = 'sample', pt.size = 0)
```

Like with the mitochondrial expression percentage, we will strive
to find a threshold that works reasonably well across all samples.
For the number of genes expressed we will want to filter out both cells
that express to *few* genes and cells that express too *many* genes. 
As noted above, damaged or dying cells may leak RNA, resulting in a low
number of genes expressed, and we want to filter out these cells to 
ignore their "damaged" transcriptomes. On the other hand, cells with way 
more genes expressed than the typical cell might be
doublets/multiplets and should also be removed.

It looks like filtering out cells that express less than 400 or 
greater than 5,000 genes is a reasonable compromise across our samples.
(Note the log scale in this plot, which is necessary for seeing the violin
densities at low numbers of genes expressed).

```{r filter_gene_counts_5k,fig.height=4,fig.width=6}
VlnPlot(liver, 'nFeature_RNA', group.by = 'sample', pt.size = 0) +
  scale_y_log10() + 
  geom_hline(yintercept = 600) + 
  geom_hline(yintercept = 5000)
#liver <- subset(liver, nFeature_RNA > 600 & nFeature_RNA < 5000)
```


### Filtering Cells by UMI

A UMI -- unique molecular identifier -- is like a molecular barcode for
each RNA molecule in the cell. UMIs are short, distinct oligonucleotides
attached during the initial 
preparation of cDNA from RNA. Therefore each UMI is unique to a single RNA
molecule. 

Why are UMI useful? The amount of RNA in a single cell is quite low 
(approximately 10-30pg according to 
[this link](https://bionumbers.hms.harvard.edu/bionumber.aspx?id=111205&ver=2&trm=amount+of+rna+per+cell&org=)).
Thus single cell transcriptomics profiling usually includes a PCR 
amplification step. PCR amplification is fairly "noisy" because small 
stochastic sampling differences can propagate through exponential 
amplification. Using UMIs, we can throw out all copies of the molecule
except one (the copies we throw out are called "PCR duplicates").

![UMI](../fig/lexogen.png)

> Note, not sure about permissions for this figure ...

Several papers (e.g. [Islam et al](https://doi.org/10.1038/nmeth.2772))
have demonstrated that UMIs reduce amplification noise in single cell
transcriptomics and thereby increase data fidelity. The only downside of
UMIs is that they cause us to throw away a lot of our data
(perhaps as high as 90% of our sequenced reads). Nevertheless, we don't
want those reads if they are not giving us new information about gene 
expression, so we tolerate this inefficiency.

CellRanger will automatically process your UMIs and the feature-barcode
matrix it produces will be free from PCR duplicates.
Thus, we can think of the number of UMIs as the sequencing depth of 
each cell.

Typically the number of genes and number of UMI are highly correlated
and this is mostly the case in our liver dataset:
```{r genes_umi}
ggplot(liver@meta.data, aes(x = nCount_RNA, y = nFeature_RNA)) +
  geom_point() +
  theme_bw(base_size = 16) +
  xlab("nUMI") + ylab("nGenes") +
  scale_x_log10() + scale_y_log10()
```




```{r filter_umi,fig.height=4,fig.width=6}
VlnPlot(liver, 'nCount_RNA', group.by = 'sample', pt.size = 0) +
  scale_y_log10() + 
  geom_hline(yintercept = 900) + 
  geom_hline(yintercept = 25000)
#liver <- subset(liver, nCount_RNA > 900 & nCount_RNA < 25000)
```

Again we try to select thresholds that remove most of the strongest outliers
in all samples. 


> ## Challenge 2
> List two technical issues that can lead to poor scRNA-seq data quality and which filters we use to detect each one. 
>
> > ## Solution to Challenge 2
> >
> > 1 ). Cell membranes may rupture during the disassociation protocol, 
which is indicated 
by high mitochondrial gene expression because the mitochondrial 
transcripts are contained within the mitochondria, while other transcripts
in the cytoplasm may leak out. Use the mitochondrial percentage filter to 
try to remove these cells.  
> > 2 ). Cells may be doublets of two different cell types. In this case
they might express many more genes than either cell type alone. Use the
"number of genes expressed" filter to try to remove these cells.  
> {: .solution}
{: .challenge}




```{r liver_outliers, include = FALSE, eval = FALSE}
# This is modeled after what the authors of the liver cell atlas did
library(scuttle)
genes <- scuttle::isOutlier(liver$nFeature_RNA, nmads = 3, 
                            batch = liver$sample, log=TRUE)
umi <- scuttle::isOutlier(liver$nCount_RNA, nmads = 3, 
                          batch = liver$sample, log=TRUE)
mt <- scuttle::isOutlier(liver$percent.mt, nmads = 3, log = FALSE,
                         batch = liver$sample, type = 'lower')
tapply(genes | umi | mt, liver$annot, mean, na.rm=T)
attr(umi, 'thresholds')
```

## Doublet detection revisited

Let's go back to our doublet predictions. How many of the cells that
are going to be filtered out of our data were predicted to be doublets
by scds?

```{r filtering}
liver$keep <- with(liver, percent.mt < 14 & nFeature_RNA > 600 &
  nFeature_RNA < 5000 & nCount_RNA > 900 & nCount_RNA < 25000)
```

Using the scds hybrid_score method, the scores range between 0 and 2.
Higher scores should be more likely to be doublets.

```{r doublet_plot}
ggplot(mutate(liver[[]], class = ifelse(keep, 'QC singlet', 'QC doublet')),
  aes(x = class, y = hybrid_score)) + 
  geom_violin() + theme_bw(base_size = 18) +
  xlab("") + ylab("SCDS hybrid score")
```

Somewhat unsatisfyingly, the scds hybrid scores aren't wildly
different between the cells we've used QC thresholds to call as doublets
vs singlets. 
There does seem to be an enrichment of cells with score >0.75 among
the QC doublets. 
If we had run scds doublet prediction on all cells we might 
compare results with *no* scds score filtering to those with an
scds score cutoff of, say, 1.0.
One characteristic of the presence of doublet cells
is a cell cluster located between two large and well-defined clusters
that expresses markers of both of them (don't worry, we will learn how
to cluster and visualize data soon). 
Returning to the scds doublet scores, we could cluster our cells with
and without doublet score filtering, and see if we note any
putative doublet clusters.

## Subset based on %MT, number of genes, and number of UMI thresholds

```{r subsetting}
liver <- subset(liver, subset = percent.mt   < 14 & 
                                nFeature_RNA > 600 &
                                nFeature_RNA < 5000 & 
                                nCount_RNA   > 900 & 
                                nCount_RNA   < 25000)
```


<!-- Discuss batch correction here? -->
<!-- it might be interesting to do batch correction across in vivo + nuc seq -->
<!-- DAS recommends using harmony if we want to do batch correction -->
<!-- Should probably do batch correction across sample -->

```{r save_seurat,include=FALSE}
# Only needed to build lessons.
# Don't use in course.
saveRDS(liver, file = file.path(data_dir, 'lesson04.rds'))
```

> ## Challenge 3
> Delete the existing counts and metadata objects. Read in the  *ex-vivo* data 
that you saved at the end of Lesson 03 (lesson03_challenge.Rdata) and create 
a Seurat object called 'liver_2'. Look at the filtering quantities and
decide whether you can use the same cell and feature filters 
that were used to create the Seurat object above.
>
> > ## Solution to Challenge 3  
> > `# Remove the existing counts and metadata.`  
> > `rm(counts, metadata)`  
> > `# Read in citeseq counts & metadata.`  
> > `load(file = file.path(data_dir, 'lesson03_challenge.Rdata'))`  
> > `# Create Seurat object.`  
> > `metadata = as.data.frame(metadata) %>%`  
> > `             column_to_rownames('cell')`  
> > `liver_2 = CreateSeuratObject(count        = counts, `  
> > `                             project      = 'liver: ex-vivo',`  
> > `                             meta.data    = metadata)`  
> > ``
> {: .solution}
{: .challenge}

> ## Challenge 4
> Estimate the proportion of mitochondrial genes. 
Create plots of the proportion of features, cells, and mitochondrial genes. 
Filter the Seurat object by mitochondrial gene expression.
>
> > ## Solution to Challenge 4  
> > `liver_2 = liver_2 %>%`  
> > `            PercentageFeatureSet(pattern = "^mt-", col.name = "percent.mt")`  
> > `VlnPlot(liver_2, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)`  
> > `liver_2 = subset(liver_2, subset = percent.mt < 10)`  
> {: .solution}
{: .challenge}


## Session Info

```{r session_info,collapse=TRUE}
sessionInfo()
```
