---
source: Rmd
title: "Quality Control of scRNA-Seq Data"
teaching: 30
exercises: 10
questions:
- "How do I determine if my single cell RNA-seq experiment data is high quality?"
- "What are the common quality control metrics that I should check in my scRNA-seq data?"
objectives:
- "Critically examine scRNA-seq data to identify potential technical issues."
- "Apply filters to remove cells that are largely poor quality/dead cells."
- "Understand the implications of different filtering steps on the data."
keypoints:
- "It is essential to filter based on criteria including mitochondrial gene expression and number of genes expressed in a cell."
- "?????"
---

```{r, setup, include=FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("04-")
```

```{r libs}
library(tidyverse)
library(Matrix)
library(Seurat)

data_dir <- '../data'
```


## Read Data from Previous Lesson

```{r load_data}
load(file.path(data_dir, 'lesson03.Rdata'))
```

## Quality control in scRNA-seq

There are many technical reasons why cells produced by an scRNA-seq protocol 
might not be of high quality. The goal of the quality control steps are to 
assure that only single, live cells are included in the final data set. 
Ultimately some multiplets and poor quality cells will likely escape your
detection and make it into your final dataset; however, these quality
control steps aim to reduce the chance of this happening.
Failure to undertake quality control is likely to adversely impact cell type 
identification, clustering, and interpretation of the data. 

Some technical questions that you might ask include:

1. Why is mitochondrial gene expression high in some cells?
1. What is UMI, and why do we check numbers of UMI?
1. What happens to make gene counts low in a cell?

<!-- DASremoved ribosomal point here. Often we don't filter based on ribo, not clearly associated with poor quality cells -->

## High-level overview of quality control and filtering

First we will walk through some of the typical quantities one 
examines when conducting quality control of scRNA-Seq data.


### Filtering Genes by Counts

As mentioned in an earlier lesson, the counts matrix is sparse and may contain 
rows (genes) or columns (cells) with low overall counts. In the case of genes, 
we likely wish to exclude genes with zeros counts in most cells. Let's see how 
many genes have zeros counts across all cells. Note that the 
[Matrix package](https://cran.r-project.org/web/packages/Matrix/index.html) 
has a special implementation of 
[rowSums](https://rdrr.io/rforge/Matrix/man/colSums.html) which works with 
sparse matrices.

```{r zero_gene_counts}
gene_counts <- Matrix::rowSums(counts, na.rm = TRUE)
sum(gene_counts == 0)
```

Of the `r nrow(counts)` genes, `r sum(gene_counts == 0)` have zero counts across 
all cells. These genes do not inform us about the mean, variance, or covariance 
of any of the other genes and we will remove them before proceeding with 
further analysis.

```{r filter_gene_by_counts}
counts <- counts[gene_counts > 0,]
```

This leaves `r nrow(counts)` genes in the counts matrix.

We could also set some other threshold for filtering genes. Perhaps we should 
look at the number of genes that have different numbers of counts. We will use 
a histogram to look at the distribution of overall gene counts. Note that, since 
we just resized the counts matrix, we need to recalculate `gene_counts`.

We will count the number of cells in which each gene was detected. Because 
`counts` is a sparse matrix, we have to be careful not to perform operations 
that would convert the entire matrix into a non-sparse matrix. This might 
happen if we wrote code like:

```{}
gene_counts <- rowSums(counts > 0)
```

The expression `counts > 0` would create a logical matrix that takes up much 
more memory than the sparse matrix. We might be tempted to try 
`rowSums(counts == 0)`, but this would also result in a non-sparse matrix 
because most of the values would be `TRUE`. However, if we evaluate 
`rowSums(counts != 0)`, then most of the values would be `FALSE`, which can be 
stored as 0 and so the matrix would still be sparse. The `Matrix` package has 
an implementation of 'rowSums()' that is efficient, but you may have to specify 
that you want to used the `Matrix` version of 'rowSums()' explicitly.

```{r gene_count_hist}
gene_counts <- Matrix::rowSums(counts > 0)

tibble(gene_id = names(gene_counts), counts = gene_counts) %>% 
  ggplot(aes(counts)) +
    geom_histogram(bins = 100) +
    labs(title = 'Number of Cells in which Gene was Detected',
         x     = 'Number of Genes',
         y     = 'Histogram of Number of Cells in which Gene was Detected') +
  theme_bw(base_size = 14) #+ scale_x_log10()

hist(gene_counts, breaks = 1000, las = 1, xlab = 'Number of Cells in which Gene was Detected', 
     ylab = 'Number of Genes', main = 'Histogram of Number of Cells in which Gene was Detected')
```

As you can see, the number of cells in which each gene is detected spans several 
orders of magnitude and this makes it difficult to interpret the plot. Some 
genes are detected in all cells while others are detected in only one cell. 
Let's zoom in on the part with lower gene counts.

```{r gene_count_hist_2}
hist(gene_counts, breaks = -1:max(gene_counts), freq = TRUE, 
     xlim = c(0, 100), las = 1, 
     xlab = 'Number of Cells in which Gene was Detected', 
     ylab = 'Number of Genes', 
     main = 'Histogram of Number of Cells in which Gene was Detected')
text(2, 1180, labels = paste(sum(gene_counts == 1), 
  'genes were detected in only one cell'), adj = c(0, 0.5))
text(3,  800, labels = paste(sum(gene_counts == 2), 
  'genes were detected in two cells'), adj = c(0, 0.5))
```

In the plot above, we can see that there are `r sum(gene_counts == 1)` 
genes that were detected in only one cell, `r sum(gene_counts == 2)` 
genes detected in two cells, etc.

Making a decision to keep or remove a gene based on its expression being
detected in a certain
number of cells can be tricky. 
If you retain all genes, you may consume more computational resources and add 
noise to your analysis. If you discard too many genes, you may miss rare 
but important cell types.

Consider this example: You have a total of 10,000 cells in your scRNA-seq 
results. There is a rare cell population consisting of 100 cells that 
expresses 20 genes which are not expressed in any other cell type. If you 
only retain genes that are detected in more than 100 cells, you will miss 
this cell population.


> ## Challenge 1
> How would filtering genes too strictly affect your results? 
How would this affect your ability to discriminate between cell types? 
>
> > ## Solution to Challenge 1
> > 
> > Filtering too strictly would make it more difficult to distinguish between
cell types. The degree to which this problem affects your analyses depends on
the degree of strictness of your filtering. Let's take the situation to its
logical extreme -- what if we keep only genes expressed in at least 95% of cells.
If we did this, we would end up with only `r sum(gene_counts > ncol(counts)*0.95)`
genes! By definition these genes will be highly expressed in all cell types,
therefore eliminating our ability to clearly distinguish between cell types.
> {: .solution}
{: .challenge}


> ## Challenge 2
> What total count threshold would you choose to filter genes? Remember that 
there are `r ncol(counts)` cells.
>
> > ## Solution to Challenge 1
> >
> > This is a question that has a somewhat subjective answer. Following from
challenge one, we do not want to be *too* strict in our filtering. However,
we do want to remove genes that will not provide much information about
gene expression variability among the cells in our dataset. Our recommendation
would be to filter genes expressed in <5 cells, but one could reasonably
justify a threshold between, say, 3 and 20 cells.
> {: .solution}
{: .challenge}


### Filtering Cells by Counts

Next we will look at the number of genes expressed in each cell. 
If a cell lyses and leaks RNA,the total number of reads in a 
cell may be low, which leads to lower gene counts. Furthermore, each
single cell suspension likely contains some amount of so-called "ambient"
RNA from damaged/dead/dying cells. This ambient RNA comes along for the ride
in every droplet. Therefore even droplets that do not contain cells
(empty droplets) can have some reads mapping to transcripts that look
like gene expression.
Filtering out these kinds of cells is a quality control step that 
should improve your final results.

We will explicitly use the `Matrix` package's implementation of 'colSums()'.

```{r sum_cell_counts}
cell_counts <- Matrix::colSums(counts > 0)
hist(cell_counts, breaks = 1000, las = 1, xlab = 'Number of Genes with Counts > 0', 
     ylab = 'Number of Cells')
```

Cells with way more genes expressed than the typical cell might be
doublets/multiplets and should also be removed.


### Creating the Seurat Object

In order to use Seurat, we must take the sample metadata and gene counts and 
create a 
[Seurat Object](https://rdrr.io/cran/SeuratObject/man/Seurat-class.html). 
This is a data structure which organizes the data and metadata and will 
store aspects of the analysis as we progress through the workshop.

Below, we will create a Seurat object for the liver data. We must first 
convert the cell metadata into a data.frame and place the barcodes 
in rownames. The we will pass the counts and metadata into the
[CreateSeuratObject](https://search.r-project.org/CRAN/refmans/SeuratObject/html/CreateSeuratObject.html) 
function to create the Seurat object. 

In the section above, we examined the counts across genes and cells and 
proposed filtering using thresholds.  The CreateSeuratObject function 
contains two arguments, 'min.cells' and 'min.features', that allow us to 
filter the genes and cells by counts. Although we might use these arguments
for convenience in a typical analysis, for this course we will look more
closely at these quantities on a per-library basis to decide on
our filtering thresholds.


```{r create_seurat_obj}
metadata = as.data.frame(metadata) %>%
              column_to_rownames('cell')
liver = CreateSeuratObject(counts       = counts, 
                           project      = 'liver: scRNA-Seq',
                           meta.data    = metadata)
```


## Typical filters for cell quality - %MT, number UMI, number of genes expressed

Here we briefly review these filters.

### Filtering by Mitochondrial Gene Content

During apoptosis, the cell membrane may break and release transcripts into 
the surrounding media. However, the mitochondrial transcripts may remain inside 
of the mitochondria. This will lead to an apparent, but spurious, increase in 
mitochondrial gene expression. As a result, we use the percentage of 
mitochondrial-encoded reads to filter out cells that were not healthy during 
sample processing. See 
[this link](https://kb.10xgenomics.com/hc/en-us/articles/360001086611) 
from 10X Genomics for additional information.

> TBD: Mention that different cell types may have different levels of mitochondrial RNA content and how we handle this.

```{r pct_mito}
liver <- liver %>% 
              PercentageFeatureSet(pattern = "^mt-", col.name = "percent.mt")
```

> TBD: Seurat uses a 5% threshold for MT content.

```{r seurat_counts_plots}
VlnPlot(liver, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```


> TBD: I made the plot below to see if there are mitochondrial expression differences by annotated cell type. The students won't have this file at this stage of the analysis. But how do we discuss these differences? Since we're looking for high values, it may not be too important.

```{r mito_by_cell_type}
liver[[c('annot', 'percent.mt')]] %>% 
    ggplot(aes(annot, percent.mt + 0.01)) +
      geom_boxplot() +
      scale_y_log10() +
      coord_flip()
```

> TBD: What threshold do we use? And provide rationale.

```{r subset_by mito}
liver <- subset(liver, subset = percent.mt < 15)
```



### Filtering Cells by Total Gene Counts

Let's look at how many genes are expressed in each cell. We'll split by
the mouse ID so we can see if there are particular samples that are 
very different from the rest.

```{r filter_gene_counts}
VlnPlot(liver, 'nFeature_RNA', group.by = 'sample')
```


> ## Challenge 2
> List three technical issues that can lead to poor scRNA-seq data quality and which filters we use to detect each one. 
>
> > ## Solution to Challenge 2
> >
> > 1 ). Cells may enter apoptosis during the disassociation protocol, which is indicated by high mitochondrial gene expression (why?).  
> > 2 ). Etc. etc. DAS to help fill in.
> {: .solution}
{: .challenge}


### Filtering Cells by UMI

> Define what UMI is. (Do we need a glossary?)
> Where do we get these from? Are they in the cell bar code?

```{r filter_umi}

```


## Cell cycle assignment 

> TBD: DAS to fill in.

## Doublet detection 

> TBD: DAS to fill in.

## Batch correction

> DAS DMG need to discuss whether to include this step.
> TBD: We will correct between in vivo, ex vivo, & nuc seq. Students will correct between in vivo & ex vivo.
> The Seurat v2 paper (Butler et al, 2018) claims to have a batch normalization method that beats ComBat & limma. Do we use that?

## Save Data for Next Lesson

We will use the Seurat object in the next lesson. Save it now and we will load it at the beginning of the next lesson. We will use R's [saveRDS](https://stat.ethz.ch/R-manual/R-devel/library/base/html/readRDS.html) command to save the Seurat object in compressed, binary format. The `saveRDS` function is useful when you want to save one object in compressed, binary format.

```{r save_seurat}
saveRDS(liver, file = file.path(data_dir, 'lesson04.rds'))
```

> ## Challenge 3
> Delete the existing counts and metadata objects. Read in the  *ex-vivo* data 
that you saved at the end of Lesson 03 (lesson03_challenge.Rdata) and create 
a Seurat object called 'liver_2'. Look at the filtering quantities and
decide whether you can use the same cell and feature filters 
that were used to create the Seurat object above.
>
> > ## Solution to Challenge 3  
> > `# Remove the existing counts and metadata.`  
> > `rm(counts, metadata)`  
> > `# Read in citeseq counts & metadata.`  
> > `load(file = file.path(data_dir, 'lesson03_challenge.Rdata'))`  
> > `# Create Seurat object.`  
> > `metadata = as.data.frame(metadata) %>%`  
> > `             column_to_rownames('cell')`  
> > `liver_2 = CreateSeuratObject(count        = counts, `  
> > `                             project      = 'liver: ex-vivo',`  
> > `                             meta.data    = metadata)`  
> > ``
> {: .solution}
{: .challenge}

> ## Challenge 4
> Estimate the proportion of mitochondrial genes. 
Create plots of the proportion of features, cells, and mitochondrial genes. 
Filter the Seurat object by mitochondrial gene expression.
>
> > ## Solution to Challenge 4  
> > `liver_2 = liver_2 %>%`  
> > `            PercentageFeatureSet(pattern = "^mt-", col.name = "percent.mt")`  
> > `VlnPlot(liver_2, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)`  
> > `liver_2 = subset(liver_2, subset = percent.mt < 10)`  
> {: .solution}
{: .challenge}


## Session Info

```{r session_info}
sessionInfo()
```
