---
source: Rmd
title: "Quality Control"
teaching: 10
exercises: 2
questions:
- "How do I deterimine if my single cell RNAseq experiment data is good?"
objectives:
- "Explain how to use RMarkdown with the new lesson template."
- "Demonstrate how to include pieces of code, figures, and challenges."
keypoints:
- "Edit the .Rmd files not the .md files"
- "Run `make serve` to knit documents and preview lesson website locally"
---

```{r, include=FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("04-")
```

## Issues in scRNA-seq

* false negatives
* higher errors for low-copy transcripts

## Typical filters for cell quality - %MT, ribosomal, number UMI, number genes 

## Differences between tissues and datatypes (single cell vs single nucleus, etc) 

## Cell cycle assignment 

## Doublet detection 

## Batch correction
