---
source: Rmd
title: "Quality Control"
teaching: 30
exercises: 10
questions:
- "How do I deterimine if my single cell RNA-seq experiment data is good?"
- "What are the common quality control metrics that I should check in my scRNA-seq data?"
objectives:
- "Critically examine scRNA-seq data to identify potential technical issues."
- "Apply filters to remove cells that may not provide high quality data."
- "Understand the implications of different filtering steps on the data."
keypoints:
- "?????"
- "?????"
---

```{r, include=FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("04-")

library(tidyverse)
library(Matrix)
library(Seurat)

data_dir = '../data'
```

## Read Data from Previous Lesson

```{r load_data}
load(file.path(data_dir, 'lesson03.Rdata'))
```

## Issues in scRNA-seq

There are many technical reasons why cells produced by an scRNA-seq protocol might not be of high quality. The goal of the quality control steps are to assure that only single, live cells are included in the final data set. Failure to do this may adversely impact cell type identification, clustering, and interpretation of the data. 

Some technical questions that you might ask include:

1. What happens to make mitochondrial expression higher?
1. Why do we check for ribosomal gene content?
1. What is UMI, what happens to make it bad, and why to we check UMI?
1. What happens to make gene counts low in a cell and how do we use gene counts per cell to fix it?

### Filtering Genes by Counts

As mentioned in an earlier lesson, the counts matrix is sparse and may contain rows (genes) or columns (cells) with low overall counts. In the case of genes, we may wish to exclude genes with zeros counts in most cells. Let's see how many genes have zeros counts across all cells. Note that the [Matrix package](https://cran.r-project.org/web/packages/Matrix/index.html) has a special implementation fo [rowSums](https://rdrr.io/rforge/Matrix/man/colSums.html) which works with sparse matrices.

```{r zero_gene_counts}
gene_counts = rowSums(counts, na.rm = TRUE)
sum(gene_counts == 0)
```

Of the `r nrow(counts)` genes, `r sum(gene_counts == 0)` have zero counts across all cells. These genes do not inform us about the mean, variance, or covariance of any of the other genes and we could remove them before proceeding with further analysis.

```{r filter_gene_by_counts}
counts = counts[gene_counts > 0,]
```

This leaves `r nrow(counts)` genes in the counts matrix.

We could also set some other threshold for filtering genes. Perhaps we should look at the number of genes that have different numbers of counts. We will use a histogram to look at the distribution of overall gene counts. Note that, since we just resized the counts matrix, we need to recalculate the gene_counts.

We will count the number of cells in which each gene was detected. Because `counts` is a sparse matrix, we have to be careful not to perform operations that would convert the entire matrix into a non-sparse matrix. This might happen if we wrote code like:

```{}
gene_counts = rowSums(counts > 0)
```

The expression `counts > 0` would create a logical matrix that takes up much more memory than the sparse matrix. We might be tempted to try `rowSums(counts == 0)`, but this would also result in a non-sparse matrix because most of the values would be `TRUE`. However, if we evaluate `rowSums(counts != 0)`, then most of the values would be `FALSE`, which can be stored as 0 and so the matrix would still be sparse. The `Matrix` package has an implementation of 'rowSums()' that is efficient, but you may have to specify that you want to used the `Matrix` version of 'rowSums()' explicitly.

```{r gene_count_hist}
gene_counts = Matrix::rowSums(counts > 0)
hist(gene_counts, breaks = 1000, las = 1, xlab = 'Number of Cells in which Gene was Detected', 
     ylab = 'Number of Genes', main = 'Histogram of Number of Cells in which Gene was Detected')
```

As you can see, the number of cells in which each gene is detected spans several orders of magnitude and this makes it difficult to interpret the plot. Some genes are detected in all cells while others are detected in only one cell. Let's zoom in on the part with lower gene counts.

```{r gene_count_hist_2}
hist(gene_counts, breaks = -1:max(gene_counts), freq = TRUE, xlim = c(0, 100), las = 1, 
     xlab = 'Number of Cells in which Gene was Detected', 
     ylab = 'Number of Genes', main = 'Histogram of Number of Cells in which Gene was Detected')
text(2, 1180, labels = paste(sum(gene_counts == 1), 'genes were detected in only one cell'), adj = c(0, 0.5))
text(3,  800, labels = paste(sum(gene_counts == 2), 'genes were detected in two cells'),     adj = c(0, 0.5))
```

In the plot above, we can see that there are `r sum(gene_counts == 1)` genes that were detected in only one cell, `r sum(gene_counts == 2)` genes detected in two cells, etc.

Selecting the number of cells in which a gene must be detected can be tricky. If you retain all genes, you may consume more computational resources and add noise to your analysis. If you discard too many genes, you may miss rare but important cells types.

Consider this example: You have a total of 10,000 cells in your scRNA-seq results. There is a rare cell population consisting of 100 cells that expresses 20 genes which are not expressed in any other cell type. If you only retain genes that are detected in more than 100 cells, you will miss this cell population.


> TDB: How many cells do I need to see a gene in for it to be informative? Discuss effects of a gene detected in 100 cells vs gene detected in 1000 cells. What are the effects of retaining too many genes?

> TBD: Can we use a plot like the one below to help explain the issues?

```{r}
ngenes = 100
ncells = 100
tmp = matrix(rbinom(ngenes * ncells, 1, prob = 5e-2), ngenes, ncells,
             dimnames = list(paste0('gene', 1:ngenes), paste0('cell', 1:ncells)))
tmp[1:10, 1:10]   = 1
tmp[11:50, 11:50] = 1
tmp[51:100,51:100]  = 1
heatmap(tmp, breaks = c(-0.5, 0.5, 10), col = c('white', 'black'))
```



> ## Challenge 1
> What total count threshold would you choose to filter genes? You may want to remake the plot above to include more cells on the X-axis. Remember that there are `r ncol(counts)` cells.
>
> > ## Solution to Challenge 1
> >
> > **TBD: Need DAS. to provide rationale. **
> {: .solution}
{: .challenge}


> ## Challenge 2
> How would filtering genes too strictly affect your results? How would this affect your ability to discriminate between cell types? 
>
> > ## Solution to Challenge 1
> >
> > **TBD: Need DAS. to provide rationale. **
> {: .solution}
{: .challenge}

### Filtering Cells by Counts

Next we will look at the number of genes expressed in each cell. If the cell processing does not go well, the total number of reads in a cell may be low, which leads to lower gene counts. Filtering out these cells is a quality control step that should improve your final results.

We will explicitly use the `Matrix` package's implementation of 'colSums()'.

```{r sum_cell_counts}
cell_counts = Matrix::colSums(counts > 0)
hist(cell_counts, breaks = 1000, las = 1, xlab = 'Number of Genes with Counts > 0', 
     ylab = 'Number of Cells')
```

> TBD: Can you have too many genes expressed? Are those doublets? 

### Creating the Seurat Object

In order to use Seurat, we must take the sample metadata and gene counts and create a [Seurat Object](https://rdrr.io/cran/SeuratObject/man/Seurat-class.html). This is a data structure which organizes the data and metadata and will store aspects of the analysis as we progress through the workshop.

Below, we will create a Seurat object for the liver data. We must first convert the cell metadata into a data.frame and place the bar codes in rownames. The we will pass the counts and metadata into the [CreateSeuratObject](https://search.r-project.org/CRAN/refmans/SeuratObject/html/CreateSeuratObject.html) function to create the Seurat object. 

In the section above, we examined the counts across genes and cells and proposed filtering by different threshold.  The CreateSeuratObject function contains two arguments, 'min.cells' and 'min.features', that allow us to filter the genes and cells by counts.

> TBD: DAS to provide rationale and values for cell and feature filtering. Also, do we want to do it here or follow the Seurat tutorial and do it after creating the object? It seems like a good idea to look at the plots before filtering. But maybe there is some basic filtering that makes sense.

```{r create_seurat_obj}
metadata = as.data.frame(metadata) %>%
              column_to_rownames('cell')
liver = CreateSeuratObject(counts       = counts, 
                           project      = 'liver: sc & nuc RNAseq',
                           meta.data    = metadata,
                           min.cells    = 100,
                           min.features = 100)
```


## Typical filters for cell quality - %MT, ribosomal, number UMI, number genes 

> Briefly review filters and reasons for them.

### Filtering by Mitochondrial Gene Content

During apoptosis, the cell membrane may break and release transcripts into the surrounding media. However, the mitochondrial transcripts will remain inside of the mitochondria. This will lead to an apparent, but spurious, increase in mitochondrial gene expression. As a result, we use the percentage of mitochondrial-encoded reads to filter out cells that were not healthy during sample processing. 

> TBD: Mention that different cell types may have different levels of mitochondrial RNA content and how we handle this.

```{r pct_mito}
liver <- liver %>% 
              PercentageFeatureSet(pattern = "^mt-", col.name = "percent.mt")
```

> TBD: Seurat uses a 5% threshold for MT content.

```{r seurat_counts_plots}
VlnPlot(liver, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```


> TBD: I made the plot below to see if there are mitochondrial expression differences by annotated cell type. The students won't have this file at this stage of the analysis. But how do we discuss these differences? Since we're looking for high values, it may not be too important.

```{r mito_by_cell_type}
liver[[c('annot', 'percent.mt')]] %>% 
    ggplot(aes(annot, percent.mt + 0.01)) +
      geom_boxplot() +
      scale_y_log10() +
      coord_flip()
```

> TBD: What threshold do we use? And provide rationale.

```{r subset_by mito}
liver <- subset(liver, subset = percent.mt < 15)
```


### Filtering Cells by Ribosomal Transcript Content

```{r filter_ribosomal}

```

### Filtering Cells by UMI

> Define what UMI is. (Do we need a glossary?)
> Where do we get these from? Are they in the cell bar code?

```{r filter_umi}

```

### Filtering Cells by Total Gene Counts

> TBD: Didn't we do this when we created the Seurat object?

```{r filter_gene_counts}

```


> ## Challenge 2
> List three technical issues that can lead to poor scRNA-seq data quality and which filters we use to detect each one. 
>
> > ## Solution to Challenge 2
> >
> > 1 ). Cells may enter apoptosis during the disassociation protocol, which is indicated by high mitochondrial gene expression (why?).  
> > 2 ). Etc. etc. DAS to help fill in.
> {: .solution}
{: .challenge}

## Differences between tissues and datatypes (single cell vs single nucleus, etc) 

> TBD: DAS to fill in.

## Cell cycle assignment 

> TBD: DAS to fill in.

## Doublet detection 

> TBD: DAS to fill in.

## Batch correction

> TBD: We will correct between in vivo, ex vivo, & nuc seq. Students will correct between in vivo & ex vivo.
> The Seurat v2 paper (Butler et al, 2018) claims to have a batch normalization method that beats ComBat & limma. Do we use that?

### Save Data for Next Lesson

We will use the Seurat object in the next lesson. Save it now and we will load it at the beginning of the next lesson. We will use R's [saveRDS](https://stat.ethz.ch/R-manual/R-devel/library/base/html/readRDS.html) command to save the Seurat object in compressed, binary format. The `saveRDS` function is useful when you want to save one object in compressed, binary format.

```{r save_seurat}
saveRDS(liver, file = file.path(data_dir, 'lesson04.rds'))
```

> ## Challenge 3
> Delete the existing counts and metadata objects. Read in the  *citeSeq* data that you saved at the end of Lesson 03 (lesson03_challenge.Rdata) and create a Seurat object called 'liver_2'. Use the same cell and feature filters that were used to create the Seurat object above.
>
> > ## Solution to Challenge 3  
> > `# Remove the existing counts and metadata.`  
> > `rm(counts, metadata)`  
> > `# Read in citeseq counts & metadata.`  
> > `load(file = file.path(data_dir, 'lesson03_challenge.Rdata'))`  
> > `# Create Seurat object.`  
> > `metadata = as.data.frame(metadata) %>%`  
> > `             column_to_rownames('cell')`  
> > `liver_2 = CreateSeuratObject(count        = counts, `  
> > `                             project      = 'liver: citeSeq',`  
> > `                             meta.data    = metadata,`  
> > `                             min.cells    = 100,`  
> > `                             min.features = 100)`  
> {: .solution}
{: .challenge}

> ## Challenge 4
> Estimate the proportion of mitochondrial genes. Create plots of the proportion of features, cells, and mitochondrial genes. Filter the Seurat object by mitochon.
>
> > ## Solution to Challenge 4  
> > `liver_2 = liver_2 %>%`  
> > `            PercentageFeatureSet(pattern = "^mt-", col.name = "percent.mt")`  
> > `VlnPlot(liver_2, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)`  
> > `liver_2 = subset(liver_2, subset = percent.mt < 10)`  
> {: .solution}
{: .challenge}

> ## Challenge 5
> TBD: Do more of the steps above.
>
> > ## Solution to Challenge 5   
> >      
> {: .solution}
{: .challenge}

### Session Info

```{r session_info}
sessionInfo()
```
