---
source: Rmd
title: "Common Analyses"
teaching: 10
exercises: 2
questions:
- "What are the most common single cell RNAseq analyses?"
objectives:
- "Explain how to use RMarkdown with the new lesson template."
- "Demonstrate how to include pieces of code, figures, and challenges."
keypoints:
- "Edit the .Rmd files not the .md files"
- "Run `make serve` to knit documents and preview lesson website locally"
---

```{r, include=FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("05-")

library(tidyverse)
library(Seurat)

data_dir = '../data'
```

## Read Data from Previous Lesson

```{r load_data}
liver = readRDS(file.path(data_dir, 'lesson04.rds'))
```


## Normalization (log and more specialized) 

### Log Normalization

> TBD: How do we show a histogram of the unnormalized and normalized values?

The count data is usually log-normally distributed. Many statistical methods work best when the data is normally distributed. 

```{r normalization, message=FALSE}
liver <- liver %>%
              NormalizeData(normalization.method = "LogNormalize")
```

### Finding Variable Features

```{r var_features, message=FALSE}
liver <- liver %>% 
              FindVariableFeatures(nfeatures = 2000)
```

### Scale Data

```{r scaling, message=FALSE}
liver <- liver %>%
              ScaleData(vars.to.regress = c("percent.mt", "nCount_RNA"))
```

### Principal Component Analysis

```{r PCA, message=FALSE}
liver <- liver %>%
              RunPCA(verbose = FALSE, npcs = 100)
```

## Dimensionality reduction (UMAP, tSNE, etc) 

> Uniform Manifold Approximation and Projection (UMAP) [van der Maaten & Hinton, 2008](https://www.jmlr.org/papers/volume9/vandermaaten08a/vandermaaten08a.pdf).
> t-Distributed Stochastic Neighbor Embedding (t-SNE) [McUnnes et al](https://arxiv.org/abs/1802.03426) 

## Clustering 

```{r seurat3, message=FALSE}
ElbowPlot(liver, ndims = 100) + geom_vline(xintercept = 27)
num_pc <- 27
liver <- FindNeighbors(liver, reduction = 'pca', 
                       dims = 1:num_pc, verbose = FALSE) %>%
           FindClusters(verbose = FALSE, resolution = 0.8) %>%
           RunUMAP(reduction = 'pca', dims = 1:num_pc, verbose = FALSE)
```

## Annotating cell types (+ automated options e.g. SingleR) 

## Finding marker genes 

## Discussion of data visualization, sample of common plots

