---
source: Rmd
title: "Common Analyses"
teaching: 10
exercises: 2
questions:
- "What are the most common single cell RNAseq analyses?"
objectives:
- "Explain how to use RMarkdown with the new lesson template."
- "Demonstrate how to include pieces of code, figures, and challenges."
keypoints:
- "Edit the .Rmd files not the .md files"
- "Run `make serve` to knit documents and preview lesson website locally"
---

```{r, include=FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("05-")
```

## Normalization (log and more specialized) 

## Dimensionality reduction (UMAP, tSNE, etc) 

## Clustering 

## Annotating cell types (+ automated options e.g. SingleR) 

## Finding marker genes 

## Discussion of data visualization, sample of common plots

