---
source: Rmd
title: "Biology Driven Analyses of scRNA-Seq"
teaching: 10
exercises: 2
questions:
- "What are some scRNA-Seq analyses that might provide me with biological insight?"
objectives:
- "Understand the ability (and limitations) of scRNA-Seq data for quantifying differences in gene expression and tissue cell composition."
- "Be able to articulate the concept of pseudotime."
- ""
- "Demonstrate how to include pieces of code, figures, and challenges."
keypoints:
- "Edit the .Rmd files not the .md files"
- "Run `make serve` to knit documents and preview lesson website locally"
---

```{r, include=FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("06-")
```

## Read Data from Previous Lesson

```{r load_data}
liver <- load(file.path(data_dir, 'lesson05.Rdata'))
```

## Identifying cell types

Let's plot the expression of some of the major cell type
markers. Look at the data.frame `markers` for a summary of the
markers we found in lesson 5.

```{r expr}
UMAPPlot(liver, label = TRUE, label.size = 6) + NoLegend()
FeaturePlot(liver, c("Lyve1", "Cdh5"), cols = c('lightgrey', 'red'))
```

## Differential expression 

Looking for differential expression can be thought of as a problem that
is related to finding marker genes.

## Pathway enrichment 

## Differences in cell composition

scCODA
linear modeling
propeller

## Cell-cell communication 

cellchat, cellphoneDB, etc.

## Pseudotime 

monocle
RNA velocity
scvelo


